x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -84,60 +97,81 @@ impl<'a, K: Key> State<'a, K> {
     /// # Errors
     /// Returns an [Error] if the key cannot be serialized or if
     /// the host fails to read the key and value.
-    /// # Panics
-    /// Panics if the value cannot be converted from i32 to usize.
     pub fn get<V>(self, key: K) -> Result<Option<V>, Error>
     where
         V: BorshDeserialize,
     {
-        #[link(wasm_import_module = "state")]
-        extern "C" {
-            #[link_name = "get"]
-            fn get_bytes(ptr: *const u8, len: usize) -> HostPtr;
+        match self.cache.borrow_mut().entry(key) {
+            Entry::Occupied(entry) => match entry.get() {
+                CachedData::Data(data) => from_slice::<V>(data)
+                    .map_err(|_| StateError::Deserialization)
+                    .map(Some),
+                CachedData::ToDelete | CachedData::Deleted => Ok(None),
+            },
+            Entry::Vacant(entry) => {
+                if let Some(bytes) = Self::get_from_host(key)? {
+                    let value = from_slice::<V>(&bytes)
+                        .map_err(|_| StateError::Deserialization)
+                        .map(Some);
+                    entry.insert(CachedData::Data(bytes));
+                    value
+                } else {
+                    entry.insert(CachedData::Deleted);
+                    Ok(None)
+                }
+            }
```

Comment:
I'm not sure how to avoid the repetition here, in one branch the data is borrowed because it comes from the storage. On the other, it has been copied from host so it's owned.

Reply:
It's working quite well though because we need an owned value to insert in the cache.

Reply:
Okay, this actually compiles
```rust
    pub fn get<V>(self, key: K) -> Result<Option<V>, Error>
    where
        V: BorshDeserialize,
    {
        #[link(wasm_import_module = "state")]
        extern "C" {
            #[link_name = "get_bytes"]
            fn get_bytes(ptr: *const u8, len: usize) -> HostPtr;
        }

        let mut cache = self.cache.borrow_mut();

        let cached_data = match cache.get(&key) {
            Some(cached_data) => cached_data,
            None => {
                let args_bytes = borsh::to_vec(&key).map_err(|_| StateError::Serialization)?;

                let ptr = unsafe { get_bytes(args_bytes.as_ptr(), args_bytes.len()) };

                if ptr.is_null() {
                    cache.insert(key, CachedData::Deleted);
                    return Ok(None);
                }

                cache.entry(key).or_insert(CachedData::Data(ptr.into()))
            }
        };

        match cached_data {
            CachedData::Data(data) => from_slice::<V>(data)
                .map_err(|_| StateError::Deserialization)
                .map(Some),
            CachedData::ToDelete | CachedData::Deleted => Ok(None),
        }
    }
```